### PR TITLE
Random insurgent units

### PR DESCRIPTION
**Pull request by veteran29**
**When merged this pull request will:**
- title

Idea based on: https://forums.bohemia.net/forums/topic/227963-random-guerilla-fighter-gear-v3/

### CODE REVIEW

**Comment by veteran29** [comment]
todo:
- [ ] prefix `CfgInsurgentsGear`
- [ ] add missionConfig support

File: addons/insurgents/XEH_preInit.sqf [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    {
+        private _key = configName _x;
+        private _items = GVAR(randomGear) getVariable _key;
+        if (isNil "_items") then {
+            LOG_1("Creating new gear category - %1",_key);
+            _items = [];
+            GVAR(randomGear) setVariable [_key, _items];
+        };
+
+        // add and filter unavilable classes
```
Comment (jonpas) [inline addons/insurgents/XEH_preInit.sqf]:
```suggestion
        // add and filter unavailable classes
```

File: addons/insurgents/XEH_preInit.sqf [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        {
+            if (!isNull (_x call CBA_fnc_getItemConfig)) then {
+                LOG_1("Adding '%1' to random gear pool",_x);
+                // we allow duplicates so chance of certain item can be increased by having them more times in arrays
+                _items pushBack _x;
+            };
+        } forEach getArray _x;
+    } forEach configProperties [_x, "isArray _x"];
```
Comment (jonpas) [inline addons/insurgents/XEH_preInit.sqf]: Slower than `configClasses`, but supports inheritance. Unsure if needed here.

Comment (veteran29) [reply addons/insurgents/XEH_preInit.sqf]: But we want properties of the class here, you can't get these with `configClasses` can you?

Comment (3Mydlo3) [reply addons/insurgents/XEH_preInit.sqf]: Wiki also says you cannon access properties with `configClasses`.
> Slightly faster than configProperties, but doesn't account for config properties or inherited entries

File: addons/insurgents/XEH_PREP.hpp [hunk 3]
```diff
@@ -0,0 +1,2 @@
+PREP(randomizeGear);
+PREP(addWeaponAmmo);
```
Comment (3Mydlo3) [inline addons/insurgents/XEH_PREP.hpp]:
```suggestion
PREP(addWeaponAmmo);
PREP(randomizeGear);
```

File: addons/insurgents/functions/fnc_randomizeGear.sqf [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+if (random 1 <= CHANCE_LAUNCHER) then {
+    private _backpack = selectRandom RANDOM_GEAR(backpacks);
+    _unit addBackpack _backpack;
+
+    private _launcher = selectRandom RANDOM_GEAR(launchers);
+    _unit addWeapon _launcher;
+
+    if (!IS_DISPOSABLE(_launcher)) then {
+        [_unit, _launcher, 3, 1] call FUNC(addWeaponAmmo);
+    };
```
Comment (3Mydlo3) [inline addons/insurgents/functions/fnc_randomizeGear.sqf]: I don't think there is any reason to add backpack for unit with disposable launcher. Although there might be a reason to add backpack for machine gunners.
```suggestion
    private _launcher = selectRandom RANDOM_GEAR(launchers);
    _unit addWeapon _launcher;

    if (!IS_DISPOSABLE(_launcher)) then {
        private _backpack = selectRandom RANDOM_GEAR(backpacks);
        _unit addBackpack _backpack;

        [_unit, _launcher, 3, 1] call FUNC(addWeaponAmmo);
    };
```

File: .vscode/settings.json [hunk 5]
```diff
@@ -9,6 +9,8 @@
         "_thisArgs",
         "_thisId",
         "_thisType",
+        "cba_disposable_replaceDisposableLauncher",
+        "cba_disposable_NormalLaunchers",
```
Comment (3Mydlo3) [inline .vscode/settings.json]: Is it really necessary? I deleted it and nothing has changed.
```suggestion
        "_thisType"
```

Comment (veteran29) [reply .vscode/settings.json]: SQFLint will warn about these and say they are undefined when not added to ignore.

Comment (3Mydlo3) [reply .vscode/settings.json]: Interesting, mine SQFLint seems to ignore them either way, even with VS restart. But if it works for you then fine. But one more thing. Shouldn't it end without `,` in line 13?

Comment (veteran29) [reply .vscode/settings.json]: VSCode has little bit more relaxed JSON format (JSON with Comments). So it works with comma at the end.

**Pull request merged**